js/ui/components/clubhouse.js

```diff
@@ -0,0 +1,323 @@
+
+    _init: function(notification) {
+        this.parent(notification);
+
+        // We don't have an "unexpanded" state for now
+        this.expanded = true;
+        this._actionBin.visible = true;
+
+        this.actor.can_focus = true;
+        this.actor.track_hover = true;
+
+        // Override the style name because this is a not a regular notification
+        this.actor.remove_style_class_name('notification-banner');
+        this.actor.add_style_class_name('clubhouse-quest-view');
+        this._iconBin.add_style_class_name('clubhouse-quest-view-icon-bin');
+
+        // Always wrap the body's text
+        this.bodyLabel.actor.clutter_text.line_wrap = true;
+        this.bodyLabel.actor.add_style_class_name('clubhouse-quest-view-label');
+
+        this._closeButton.connect('clicked', Lang.bind(this, function() {
+            this.emit('dismiss');
+        }));
+    },
+});
+Signals.addSignalMethods(ClubhouseNotificationBanner.prototype);
+
+var ClubhouseNotification = new Lang.Class({
+    Name: 'ClubhouseNotification',
+    Extends: NotificationDaemon.GtkNotificationDaemonNotification,
+
+    _init: function(source, notification) {
+        this._currentIconPath = null;
+
+        let gicon = notification.icon;
+        if (gicon) {
+            let newGicon = this._getClubhouseIcon(Gio.icon_deserialize(gicon));
+            notification.icon = newGicon.serialize();
+        }
+
+        // We don't set up a title for now
+        notification.title = GLib.Variant.new('s', '');
+
+        this.parent(source, notification);
+    },
+
+    _getClubhouseIcon: function(gicon) {
+        let iconPath = gicon.to_string();
+
+        if (iconPath.startsWith('/app/'))
+            iconPath = iconPath.substring(5);
+
+        // Check if this icon is already loaded, in which case we return null so it won't
+        // be updated.
+        if (this._currentIconPath == iconPath)
+            return null;
+
+        this._currentIconPath = iconPath;
+        let iconFile = this._getClubhouseAsset(iconPath);
+
+        return Gio.FileIcon.new(iconFile);
+    },
+
+    _getClubhouseAsset: function(assetSubpath) {
+        let installationPath = Flatpak.Installation.new_user(null).get_path().get_path();
+        let assetPath = GLib.build_filenamev([installationPath, 'app', CLUBHOUSE_ID, 'current',
+                                               'active', 'files', assetSubpath]);
+        let assetFile = Gio.File.new_for_path(assetPath);
+
+        if (assetFile.query_exists(null))
+            return assetFile;
+
+        return null;
+    },
+
+    createBanner: function() {
+        return new ClubhouseNotificationBanner(this);
+    },
+
+    updateFromParams: function(params) {
+        // Update actions/buttons
+        this.actions = [];
+
+        if (this._buttonBox) {
+            this._buttonBox.destroy();
+            this._buttonBox = null;
+        }
+
+        let buttons = params.buttons;
```

Comment:
This is the `params.buttons` given for the notification, and since I am reusing an action, I still had to reset/clear actions when updating it.

Reply:
I see, I had missed that code in `notificationDaemon.js`. I feel that the way this is setup is going against how GtkNotificationDaemon notifications are supposed to work.
In other words, I'd like this code to run as part of the regular constructor of `GtkNotificationDaemonNotification` instead of here. More on that below.

Reply:
Okay. I have no time to change all that implies using the original code. I am happy if you want to change it, hopefully you will see my point better, or I will see yours :smile: